### ISSUE TITLE
Go to definition sometimes goes to the declaration instead

### ISSUE THREAD

**Issue by sean-mcmanus** [microsoft/vscode-cpptools#6254]
I am working in VSCode Version: 1.49.3 and have this bug, macOS 10.15.4

When I click "Go to definition", sometimes it went to declarations and sometimes it went to the definition.
No error or warning messages.

c_cpp_properties.json

```
{
  "configurations": [
    {
      "name": "vscode_stm32_c_cpp_properties",
      "compilerPath": "/Users/AneoX/compilers/gcc-arm-none-eabi-9-2020-q2-update/bin/arm-none-eabi-gcc",
      "includePath": [
        "${workspaceRoot}/Inc/",
        "${workspaceRoot}/Drivers/cmsis/",
        "${workspaceRoot}/Drivers/STM32F0xx_StdPeriph_Driver/inc/"
        // "${workspaceRoot}/**"
      ],
      "browse": {
        "path": ["${workspaceFolder}"],
        "limitSymbolsToIncludedHeaders": true,
        "databaseFilename": "${workspaceFolder}/.vscode/browse.vc.db"
      },
      "cStandard": "c11",
      "cppStandard": "c++17",
      "intelliSenseMode": "gcc-arm",
      "configurationProvider": "ms-vscode.cmake-tools",
      "defines": ["USE_STDPERIPH_DRIVER", "STM32F042"]
    }
  ],
  "version": 4
}
```

_Originally posted by @aneox in https://github.com/microsoft/vscode-cpptools/issues/3378#issuecomment-703267452_

**Comment by sean-mcmanus** [comment]
Go to definition going to the declaration instead of the definition is a symptom, which could have multiple root causes. Are you able to provide an isolated repro example? It's possible the IntelliSense type checking is failing when comparing the definition and the source identifier.

**Comment by aneox** [comment]
@sean-mcmanus please take a look, 
https://github.com/aneox/vscode_stm32f042_template
https://youtu.be/lFwKBK_vSsA
In video you can see, if I click "goto definitions", it went to the declaration. 
But if I manually find src file, opened and closed it, then "goto to definition" become workable.

**Comment by sean-mcmanus** [comment]
I'm not reproing the issue. Can you do a C/C++: Reset IntelliSense Database command and look for stm32f0xx_rcc.c in the C/C++ logging (with C_Cpp.loggingLevel set to Debug)? It looks like something is causing that file to not be parsed on your system.

You should see something like
```
------- Workspace parsing diagnostics -------
Number of folders and files enumerated: 506
Number of files discovered (not excluded): 251
Number of files parsed: 99
```

**Comment by aneox** [comment]
I do not find where to see this stat
> ```
> ------- Workspace parsing diagnostics -------
> Number of folders and files enumerated: 506
> Number of files discovered (not excluded): 251
> Number of files parsed: 99
> ```

But I made Reset IntelliSense Database and set debug level, no error messages and the same bug occur

**Comment by sean-mcmanus** [comment]
The stats are from running C/C++: Log Diagnostics.

Did you see stm32f0xx_rcc.c in the "C/C++" logs? It should say "Tag parsing: ...stm32f0xx_rcc.c".

**Comment by aneox** [comment]
> The stats are from running C/C++: Log Diagnostics.

no info in C/C++: Log Diagnostics, only configuration print


> Did you see stm32f0xx_rcc.c in the "C/C++" logs? It should say "Tag parsing: ...stm32f0xx_rcc.c".

in this log i see info only about the opened file, no info about parsing. How i should activate workspace parsing?

**Comment by aneox** [comment]
C/C++ log print after reset, searching for files..., files - 0

**Comment by sean-mcmanus** [comment]
What version of the C/C++ extension are you using?

**Comment by aneox** [comment]
v1.0.1
when I hit save in the config file, then I see some output in С/С++ about parsing, it try to parse only *.h files

**Comment by aneox** [comment]
> I'm not reproing the issue. Can you do a C/C++: Reset IntelliSense Database command and look for stm32f0xx_rcc.c...

Actually bug occurs not only with stm32f0xx_rcc.c, with all files.

**Comment by aneox** [comment]
@sean-mcmanus I have installed VScode on Windows in the virtual machine on my Mac, "GoTo definition" works perfectly.

But in Mac version, do not work, so boring. 
I want to switch my workspace from QT Creator ide to Code, but this bug so boring, please help.

**Comment by aneox** [comment]
@sean-mcmanus ok, looks like I find a reason, after removing the git folder, bug gone
`cd workspace && rm -rf .git`
how I can exclude .git folder from parsing?

**Comment by aneox** [comment]
I have added exclude git in settings.json with db reset
```
  "files.exclude": {
    "**/.git": true,
    "**/build": true
  },
  "search.exclude": {
    "**/.git": true,
    "**/build": true
  }
```

the same bug

**Comment by sean-mcmanus** [comment]
I wasn't able to repro this on Mac either. @Colengms @michelleangela Can either of you repro it?

It sounds like it could be an issue with "configurationProvider": "ms-vscode.cmake-tools" being set. That can override the browse.path setting. You might want to try removing that or looking to see if the logging indicates it's sending an invalid value for that.

**Comment by aneox** [comment]
@sean-mcmanus yes, after i removed "configurationProvider": "ms-vscode.cmake-tools", bug gone, works great

this line has been added with copy-paste from some tutorial.
Thank you!

**Comment by aneox** [comment]
@sean-mcmanus maybe I do not need to remove this line, but find a fix in CMake config?

**Comment by sean-mcmanus** [comment]
@aneox Yeah, for some reason the CMake Tools extension is not sending the correct browse info. I'm moving this issue to that extension.
